vllm/model_executor/models/chatglm.py

```diff
@@ -0,0 +1,557 @@
+                    self.num_multi_query_groups_per_partition * self.hidden_size_per_attention_head,
+                    self.num_multi_query_groups_per_partition * self.hidden_size_per_attention_head,
+                ],
+                dim=-1,
+            )
+            query_layer = query_layer.contiguous().view(
+                query_layer.size()[:-1] + (self.num_attention_heads_per_partition, self.hidden_size_per_attention_head)
+            )
+            key_layer = key_layer.contiguous().view(
+                key_layer.size()[:-1] + (self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+            )
+            value_layer = value_layer.contiguous().view(
+                value_layer.size()[:-1]
+                + (self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+            )
+        else:
+            new_tensor_shape = mixed_x_layer.size()[:-1] + \
+                               (self.num_attention_heads_per_partition,
+                                3 * self.hidden_size_per_attention_head)
+            mixed_x_layer = mixed_x_layer.contiguous().view(*new_tensor_shape)
+
+            # [sq, b, np, 3 * hn] --> 3 [sq, b, np, hn]
+            (query_layer, key_layer, value_layer) = split_tensor_along_last_dim(mixed_x_layer, 3)
+
+        k_cache, v_cache = kv_cache
+        rotary_pos = build_rotary_pos(self.seq_length, self.rotary_dim // 2, dtype=query_layer.dtype, device=query_layer.device, rope_ratio=self.rope_ratio)
+        if positions is not None:
+            rotary_pos = rotary_pos[positions]
+        else:
+            rotary_pos = rotary_pos[:query_layer.shape[0]]
+
+        rotary_pos = rotary_pos.unsqueeze(0).transpose(0, 1).contiguous()
+
+        query_layer = apply_rotary_pos_emb(query_layer, rotary_pos).reshape(-1, self.num_attention_heads_per_partition, self.hidden_size_per_attention_head)
+        key_layer = apply_rotary_pos_emb(key_layer, rotary_pos).reshape(-1, self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+        
+
+        if self.multi_query_attention:
+            key_layer = key_layer.unsqueeze(-2)
+            key_layer = key_layer.expand(
+                -1, -1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, -1
+            )
+            key_layer = key_layer.contiguous().view(
+                key_layer.size()[:1] + (self.num_attention_heads_per_partition, self.hidden_size_per_attention_head)
+            )
+            value_layer = value_layer.unsqueeze(-2)
+            value_layer = value_layer.expand(
+                -1, -1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, -1
+            )
+            value_layer = value_layer.contiguous().view(
+                value_layer.size()[:1] + (self.num_attention_heads_per_partition, self.hidden_size_per_attention_head)
+            )
```

Comment:
vLLM has native support for multi-query attention and this `expand` operation can be avoided. This should greatly improve the speed and memory utilization of ChatGLM models. You can refer to GPTBigCode's implementation below: https://github.com/vllm-project/vllm/blob/a57d13cc967d04478cebcfa2c3440fb167b21c6d/vllm/model_executor/models/gpt_bigcode.py#L59-L89

Reply:
Hi, can I just replace `expand` operation with `repeat_interleave` ? I found if I passed the `num_kv_heads` (which is 2 in chatglm2) params to `PagedAttention`, the output would be confusing. I thought it was caused by the `self.head_mapping`. If you have some other advice, please tell me.

Reply:
Hi @canghongjian! Sorry for the delayed response. The current implementation uses multi-head attention kernel instead of the grouped-query attention in vLLM. Using the multi-head attention kernel requires a redundant copy of key and value tensors. In other words, the `expand` or the `repeat_interleave` operators here are redundant and waste memory.

I believe setting `num_kv_heads=2` is correct, but it seems like the default `head_mapping` is incorrect. I took a look at the [chatGLM's source code on huggingface](https://huggingface.co/THUDM/chatglm2-6b/blob/b1502f4f75c71499a3d566b14463edd62620ce9f/modeling_chatglm.py#L419-L433), it seems like instead of letting the `head_mapping` to be:

`[0, 0, ..., 0, 0, 1, 1, ..., 1, 1]`

It should be:

`[0, 1, ..., 0, 1, 0, 1, ..., 0, 1]`.

Can you try to modify the `torch.repeat_interleave` to `torch.repeat` in `attention.py` (below) and see whether this can make the model correct?

https://github.com/vllm-project/vllm/blob/65fc1c3127f4df8a99a0a5db4a205dbfcc673713/vllm/model_executor/layers/attention.py#L69-L71

https://github.com/vllm-project/vllm/blob/65fc1c3127f4df8a99a0a5db4a205dbfcc673713/vllm/model_executor/layers/attention.py#L105-L108

Reply:
I have tried @zhuohan123 . The main modification is:
```
# in chatglm.py
......

self.atten = PagedAttention(config.num_attention_heads,
                            self.hidden_size_per_attention_head,
                            self.norm_factor,
                            num_kv_heads=(self.num_multi_query_groups_per_partition
    if self.multi_query_attention else config.num_attention_heads))

......

query_layer = apply_rotary_pos_emb(query_layer, rotary_pos).reshape(
    -1, (self.num_attention_heads_per_partition *
    self.hidden_size_per_attention_head))
key_layer = apply_rotary_pos_emb(key_layer, rotary_pos).reshape(
    -1, (self.num_multi_query_groups_per_partition *
    self.hidden_size_per_attention_head))

context_layer = self.atten(query_layer, key_layer, value_layer,
                                   k_cache, v_cache, input_metadata,
                                   cache_event)
```

```
# in attention.py
self.head_mapping = torch.arange(self.num_kv_heads, dtype=torch.int32, device="cuda").repeat(self.num_queries_per_kv)

......

if self.num_kv_heads != self.num_heads:
    # Project the key and value tensors to the desired number of heads.
    # key = torch.repeat_interleave(key, self.num_queries_per_kv, dim=1)
    # value = torch.repeat_interleave(value,
    #                                 self.num_queries_per_kv,
    #                                 dim=1)
    key = key.repeat([1,self.num_queries_per_kv,1])
    value = value.repeat([1,self.num_queries_per_kv,1])
```
However, the output is still confusing. Did I make mistakes?

In addition, I don't really understand your words "Using the multi-head attention kernel requires a redundant copy of key and value tensors." In my original code, I just copy the key and value tensors before `PagedAttention` call, and in `attention.py` it will not step into the `key = torch.repeat_interleave` part because the `self.num_kv_heads` is equal to `self.num_heads` when I don't pass the `num_kv_heads` param. Therefore, the copy operation is only performed once in both settings.

Looking forward to your reply.

Reply:
Oh, I believe my main point is that our `PagedAttention` class supports the case where the number of query heads is not the same as the number of key/value heads. In your code, this property is not utilized. In other words, "the copy of the key and value tensors before PagedAttention" is redundant and suppose to be avoided.

I believe the confusing results are caused by the incorrect mapping between query heads and key/value heads. When the number of query heads is different from key/value heads, we use the `head_mapping` array to record which key/value head each query head corresponds to.

Can you take a more detailed look into this mapping? I believe once we set the `head_mapping` correct, we should be able to get correct results. Let me know if there are any difficulties.

Reply:
After some attempts, I still failed. Could you please try some debugging?